### PR TITLE
[#30] Semantic / Smart Selection

### PR DESCRIPTION
**Pull request by xyin96**
I'm very new to Rust, so if there is a better way to do this, let me know!

This PR is to implement semantic (or smart) selection through double, or triple click events (#30).

TODOs:
 - [x] Implement Double Click handler
 - [x] Implement Triple Click handler
 - [x] Implement Semantic selection
 - [x] Implement `BidrectionalIterator`
 - [x] Add configurable definitions for selection stopping characters
 - [x] Add configurable double (and triple) click threshold
 - [x] Add unit tests for semantic selection
 - [x] Add unit tests for double / triple click
 - [x] Add unit tests for `BidirectionalIterator` / `GridIterator`

### CODE REVIEW

File: src/term/mod.rs [hunk 1]
```diff
@@ -419,6 +419,54 @@ impl Term {
         self.dirty
     }
 
+    pub fn semantic_selection(&self, selection: &mut Selection, point: Point) {
+        let escape_chars = ['│', '`', '|', ':', '"', '\'', ' ', '(', ')', '[', ']', '{', '}', '<', '>'];
```
Comment (kurnevsky) [inline src/term/mod.rs]: Maybe make it const?

Comment (xyin96) [inline src/term/mod.rs]: My plan was going to make this configurable through the config file

**Comment by jwilm** [comment]
Thanks for working on this! The plans for config and tests sounds great.

One comment on implementation at this time--implementing a [`DoubleEndedIterator`] for the `Grid<T>` would allow the code to be cleaned up significantly. All of the manual indexing can be moved into a reusable component, and your code can just apply the selection logic. I'd be happy to help out with this part if you like.

**Comment by xyin96** [comment]
Thanks for the feedback!

I did some research on `DoubleEndedIterator` , and (feel free to correct me if I'm missing something) it doesn't look like it will do what we want it to do since the behavior I think we're looking for the pointers to iterate outwards from the selected `Point`. It seems that a `DoubleEndedIterator` is not designed to have the pointers intersect.

I do agree that moving the indexing logic into some sort of `Iterator` is the way to go though.

**Comment by jwilm** [comment]
d'oh, for some reason I was thinking it was `BidirectionalIterator`. I don't think there's actually a trait defined for that, but maybe we can do something similar anyway. Sorry for the confusion.

Something like this should work:

```rust
trait BidirectionalIterator: Iterator {
    fn prev(&mut self) -> Option<Self::Item>;
}
```

I've put together [an example] to make sure borrowing works out. Naturally, the grid implementation would be a bit more complicated.

File: alacritty_macos.yml [hunk 2]
```diff
@@ -205,6 +205,28 @@ key_bindings:
 mouse_bindings:
   - { mouse: Middle, action: PasteSelection }
 
+multiclick_threshold: 300
+
+semantic_escape_chars: [
```
Comment (jwilm) [inline alacritty_macos.yml]: Can these just be specified in a big string like `",|:\"'()[]}{><"`?

File: src/term/mod.rs [hunk 3]
```diff
@@ -460,7 +463,7 @@ impl Term {
 
         while {
```
Comment (jwilm) [inline src/term/mod.rs]: Could shorten this to

```rust
while let Some(cell) = left_iter.prev() {
    if !self.semantic_escape_chars.contains(&cell.c) {
        // update side_left
    }
}
```

File: src/term/mod.rs [hunk 4]
```diff
@@ -472,7 +475,7 @@ impl Term {
         let mut right_iter = self.grid.iter_from(point);
         while {
```
Comment (jwilm) [inline src/term/mod.rs]: ditto

**Comment by jwilm** [comment]
Love watching the progress here! The iterator-based solution is so much cleaner.

**Comment by xyin96** [comment]
Thanks for the review, @jwilm 

I was working on using a `String` for the config, but ran into an issue when I tried to do a `escape_chars.contains(&cell.c)`. The compiler suggests that `FnOnce` is not implemented on `char` and that I should implement it. What is your stance on this? Alternatively I could try casting the `char` to a `string` since that seems to work already, but feels a bit hacky.

**Comment by jwilm** [comment]
I think that error should go away if you pass `cell.c` instead of `&cell.c`.

File: src/term/mod.rs (hunk 3 above)
Comment (xyin96) [inline src/term/mod.rs]: I opted to keep the `!self.semantic_escape_chars.contains(...)` in the condition because there are two cases where we want to stop iteration:
 - There is nothing left to iterate over in that direction
 - Cell contains an escape character

**Comment by xyin96** [comment]
I think that should be feature completion. I think we just need to add some test coverage for:
 - Multiclick handlers
 - Semantic Selection Logic
 - Bidirectional Iterator (and the Grid implementation of it)

**Comment by xyin96** [comment]
Not quite sure how to add tests for the DoubleClick, but other than that, I added some basic coverage for both `GridIterator` and `semantic_selection`.

**Comment by jwilm** [comment]
@xyin96 if you manually create the event processor and terminal, you could manually feed it click events and then check the selection state.

**Comment by xyin96** [comment]
I occasionally get some test sporadic test failures. 

My hypothesis is that this can be resolved if we can do some headless testing or stub out the display, and not having to start up a `glutin::Window` for each test.

Let me know what you think.

**Comment by jwilm** [comment]
Hmm, it seems that the input processing is extremely coupled to everything else. I'll push a patch to fix that in the morning.

After that's fixed, the input processing will be *much* easier to test, and you won't need to create a glutin window or half of that other stuff. Sorry for the mess! 😳

**Comment by xyin96** [comment]
Sounds good, thanks!

**Comment by xyin96** [comment]
Tests seem to be working now. Let me know what you think.

**Comment by jwilm** [comment]
Ah cool, ready for a review?

File: alacritty_macos.yml [hunk 5]
```diff
@@ -205,6 +205,10 @@ key_bindings:
 mouse_bindings:
   - { mouse: Middle, action: PasteSelection }
 
+multiclick_threshold: 300
```
Comment (jwilm) [inline alacritty_macos.yml]: The units should be part of the name here to avoid ambiguity. Eg: `multiclick_threshold_ms`

File: src/config.rs [hunk 6]
```diff
@@ -280,6 +290,8 @@ impl Default for Config {
             colors: Default::default(),
             key_bindings: Vec::new(),
             mouse_bindings: Vec::new(),
+            semantic_escape_chars: String::new(),
+            multiclick_threshold: 1000,
```
Comment (jwilm) [inline src/config.rs]: Should probably be 300 like elsewhere

File: src/event.rs [hunk 7]
```diff
@@ -20,6 +20,7 @@ use term::{Term, SizeInfo, TermMode};
 use util::limit;
 use util::fmt::Red;
 use window::Window;
+use time::{self, Timespec};
```
Comment (jwilm) [inline src/event.rs]: I think `std::time::{Instant, Duration}` should be sufficient for the usage here, and we don't need a third-party dependency. Recording click time would be

```rust
Instant::now()
```

Comparing click times would be

```rust
let now = Instant::now();
let delta /* Duration */ = now.duration_since(last_click_time);
if delta < click_threshold {
    // ..
}
```

File: src/event.rs [hunk 8]
```diff
@@ -156,7 +178,7 @@ impl<N: Notify> Processor<N> {
     /// Handle events from glutin
     ///
     /// Doesn't take self mutably due to borrow checking. Kinda uggo but w/e.
-    fn handle_event<'a>(
+    pub fn handle_event<'a>(
```
Comment (jwilm) [inline src/event.rs]: I don't think this should be `pub` just for tests. Tests for this system can just be added in the same file.

File: src/grid.rs [hunk 9]
```diff
@@ -193,6 +210,53 @@ impl<T> Grid<T> {
...
 }
 
+impl<'a, T> Iterator for GridIterator<'a, T> {
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let last_line = self.grid.num_lines() - Line(1);
+        let last_col = self.grid.num_cols() - Column(1);
+        match self.cur {
+            Point { line, col } if
```
Comment (jwilm) [inline src/grid.rs]: I think this match should be refactored to not have the `if` blocks here. Not sure if we can match directly on struct patterns, but should be possible to destructure and match on the tuple values. Eg:

```rust
match (line, col) {
    (Line(0), Column(0)) => /* stuff */,
    // etc
}
```

File: src/grid.rs [hunk 10]
```diff
@@ -588,4 +652,51 @@ mod tests {
...
+        assert_eq!(Column(4), iter.cur.col);
+        assert_eq!(Line(0), iter.cur.line);
+
+
+        // test that iter ends at end of grid
+        let mut final_iter = grid.iter_from(Point {
+            line: Line(4),
+            col: Column(4),
+        });
+        assert_eq!(None, final_iter.next());
```
Comment (jwilm) [inline src/grid.rs]: Should check that `prev()` works at the end.

File: src/input.rs [hunk 11]
```diff
@@ -324,7 +338,23 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
...
+                        let diff = time::get_time() - self.ctx.mouse_mut().last_click_timestamp;
+                        let within_threshold = &diff.num_milliseconds() < self.multiclick_threshold;
+                        match self.ctx.mouse_mut().click_state {
+                            ClickState::Click if within_threshold => {
+                                self.ctx.mouse_mut().click_state = ClickState::DblClick;
+                                self.on_mouse_dbl_click();
+                            },
+                            ClickState::DblClick if within_threshold => {
+                                self.ctx.mouse_mut().click_state = ClickState::TripleClick;
+                                println!("Triple Click!");
```
Comment (jwilm) [inline src/input.rs]: Why not do something with triple click? At the very least, this should have a big `// TODO` and change the `println!` to a `debug!`.

File: src/input.rs [hunk 12]
```diff
@@ -410,14 +440,79 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
...
+                    height: 51.0,
+                    cell_width: 3.0,
+                    cell_height: 3.0,
+                };
+
+                let terminal = Term::new(&config, size);
+                let terminal = Arc::new(FairMutex::new(terminal));
+
+                let (tx, ..) = mpsc::channel();
+                let (sender, ..) = mio::channel::channel();
```
Comment (jwilm) [inline src/input.rs]: You shouldn't need all of this setup anymore. If the point is just to test input processing, we should have an `ActionContext` implementation just for the tests.

File: src/input.rs [hunk 13]
```diff
@@ -437,6 +532,27 @@ mod tests {
...
+        initial_state: ClickState::None,
+        input: Event::MouseInput(ElementState::Pressed, MouseButton::Left),
+        end_state: ClickState::Click
+    }
+
+    test_clickstate! {
+        name: double_click,
+        initial_state: ClickState::Click,
+        input: Event::MouseInput(ElementState::Pressed, MouseButton::Left),
+        end_state: ClickState::DblClick
```
Comment (jwilm) [inline src/input.rs]: I would expect the tests to check that the handler methods are called as appropriate as well.

File: src/term/mod.rs [hunk 14]
```diff
@@ -436,18 +438,53 @@ impl Term {
...
+            col: point.col
+        };
+        let mut side_right = Point {
+            line: point.line,
+            col: point.col
+        };
+
+        let mut left_iter = self.grid.iter_from(point);
+
+        while match left_iter.prev() {
```
Comment (jwilm) [inline src/term/mod.rs]: Should use

```rust
while let Some(cell) = left_iter.prev() {
    if self.semantic_escape_chars.contains(cell.c) {
        break;
    }
    // update side_left
}
```

File: src/term/mod.rs [hunk 15]
```diff
@@ -436,18 +438,53 @@ impl Term {
...
+        self.custom_cursor_colors = config.custom_cursor_colors();
+        self.semantic_escape_chars = config.semantic_escape_chars().clone();
     }
 
     #[inline]
     pub fn needs_draw(&self) -> bool {
         self.dirty
     }
 
+    pub fn semantic_selection(&self, selection: &mut Selection, point: Point) {
```
Comment (jwilm) [inline src/term/mod.rs]: Seems like this doesn't handle triple clicking at all. Is that intended to use a separate function?

File: src/term/mod.rs [hunk 16]
```diff
@@ -1271,6 +1308,38 @@ mod tests {
...
+
+        let mut escape_chars = String::from("\"");
+
+        mem::swap(&mut term.grid, &mut grid);
+        mem::swap(&mut term.semantic_escape_chars, &mut escape_chars);
+
+        let mut selection = Selection::new();
+        term.semantic_selection(&mut selection, Point { line: Line(3), col: Column(0) });
+        match selection.span() {
+            Some(span) => assert_eq!(term.string_from_selection(&span), "a\na\na"),
```
Comment (jwilm) [inline src/term/mod.rs]: Hmm, should double click span multiple lines? I was under the impression that double click would only ever match within a line and that triple click should expand across multiple lines *if* the terminal had wrapped a long line.

Basically, I guess we didn't talk about how semantic selection could behave. We should embrace the principal of least surprise here and attempt to match behavior from other terminal emulators.

File: src/grid.rs (hunk 9 above)
Comment (xyin96) [inline src/grid.rs]: So I did some research on this, it doesn't look like we can escape the `if` blocks here because that is variable. It looks like under the hood, Rust `match`es patterns are static. You can do ranges, but it freaks out when you try to give it a variable to for the range bounds.

File: src/input.rs (hunk 13 above)
Comment (xyin96) [inline src/input.rs]: Handler methods are in the same block as the `MouseState` state bump. Should I move those state transitions into the handler functions to ensure that the handler functions are being called?

**Comment by xyin96** [comment]
I've addressed most of the changes that you've requested, and left a few notes on the others. Let me know what you think.

**Comment by jwilm** [comment]
Running short on time right now; I'll try and review later today. Worst case, I'll get to it in the morning.

File: src/grid.rs (hunk 9 above)
Comment (jwilm) [inline src/grid.rs]: Oh interesting. I would've expected it to handle dynamic values. Thanks for explaining!

File: src/term/mod.rs [hunk 17]
```diff
@@ -436,18 +438,70 @@ impl Term {
...
+
+        while let Some(cell) = left_iter.prev() {
+            if self.semantic_escape_chars.contains(cell.c) {
+                break;
+            }
+            side_left.col = left_iter.cur.col;
+            side_left.line = left_iter.cur.line;
+
+            if left_iter.cur.col == Column(0) {
+                break; // cut off at beginning of line
```
Comment (jwilm) [inline src/term/mod.rs]: I did some testing in iTerm, and it looks like double click will actually work across lines *if* it detects that the line has wrapped. You can test this by echoing a URL that spans lines.

To accomplish the same thing in Alacritty, we would decide to continue expanding the selection when `iter.cur.col == grid.num_cols() - 1` and `cell.flags.contains(flags::WRAPLINE)`.

This case is distinct from when there is a linefeed (eg `\n`) which should break the selection. I think that behavior should fall out of the check described above.

File: src/input.rs (hunk 13 above)
Comment (jwilm) [inline src/input.rs]: I think the state transitions should still be done where they are (the processor is solely responsible for updating the Mouse state). The `ActionContext` trait advertises that the methods will be called though, so it would be nice to have coverage on that as well.

File: alacritty.yml [hunk 18]
```diff
@@ -205,6 +205,9 @@ key_bindings:
 mouse_bindings:
   - { mouse: Middle, action: PasteSelection }
 
+multiclick_threshold_ms: 300
```
Comment (jwilm) [inline alacritty.yml]: What do you think about putting these new properties into a struct? Maybe something like

```rust
selection:
  multiclick_threshold_ms: 300
  escape_characters: ",│`|:\"' ()[]{}<>"
```

Comment (xyin96) [inline alacritty.yml]: I think `multiclick_threshold_ms` should probably put under `mouse` config if anything.

I can put the `escape_chars` under `selection`, but it wouldn't make much sense right now since it's the only field, (unless we want to make selection more customizable).

File: src/term/mod.rs (hunk 17 above)
Comment (xyin96) [inline src/term/mod.rs]: So I've implemented (and added test coverage for) this, but is it a known issue that this feature doesn't work in `tmux`?

File: alacritty.yml (hunk 18 above)
Comment (jwilm) [inline alacritty.yml]: You've got a point about the mouse config. I would like to keep breaking config changes to a minimum, so it'd be nice to ship something that could potentially be the final arrangement. Maybe something like this:

```yaml
# Configuration for the mouse
# Multi-click thresholds are in units of milliseconds.
mouse:
  double_click: { threshold: 300, action: SelectSemantically }
  triple_click: { threshold: 300, action: SelectLine }

selection:
  semantic_escape_chars: ",│`|:\"' ()[]{}<>"
```

The same config type (`Multiclick`, for example) could be used for both double and triple click sections. I think it would be nice to have a block for `selection` even if there's only one property initially.

We could probably omit the `action` field of the `Multiclick` config for now since making them into an action might be kind of annoying. Adding it later would be backwards compatible with this layout.

I realize this change might be sort of onerous. If you're getting sick of applying these changes, I'd be happy to push commits on top. Just let me know :).

Comment (xyin96) [inline alacritty.yml]: That sounds reasonable -- I can jump on it as soon as I have time (which will probably be sometime **tomorrow** afternoon). If you would like to manually jump in and apply the changes yourself to close this PR sooner, I am fine with that too.

**Comment by xyin96** [comment]
I pushed a few more commits regarding your review. I'm not sure if there is a way to directly test whether a function has been called or not, but I have a working alternative for that. WDYT?

**Comment by jwilm** [comment]
Rather than do another round of review, I pushed a commit with the changes I had in mind. I think it's good to go.

Assuming you don't see any problems, would you mind squashing? You're welcome to just collapse my changes into your commits.

**Comment by xyin96** [comment]
Squashed and ready to go!

**Comment by jwilm** [comment]
Thanks! Going to wait until travis is green and then I'll merge.
